jcl/src/java.base/share/classes/openj9/internal/tools/attach/target/DiagnosticUtils.java

```diff
@@ -348,6 +387,128 @@ private static DiagnosticProperties doDump(String diagnosticCommand) {
+
+		// convert to milliseconds based on the unit
+		switch (timeUnit) {
+		case "s":
+			timeInMilli = TimeUnit.SECONDS.toMillis(time);
+			break;
+		case "m":
+			timeInMilli = TimeUnit.MINUTES.toMillis(time);
+			break;
+		case "h":
+			timeInMilli = TimeUnit.HOURS.toMillis(time);
+			break;
+		case "d":
+			timeInMilli = TimeUnit.DAYS.toMillis(time);
+			break;
+		default:
+			// no unit or unrecognized unit, assume milliseconds
+			timeInMilli = time;
+			break;
+		}
+		return Long.valueOf(timeInMilli);
+	}
+
+	private static Long parseTimeParameter(String paramName, String[] parameters) {
+		for (String param : parameters) {
+			if (param.startsWith(paramName + "=")) {
+				int valueStart = param.indexOf("=");
+				if (valueStart != -1) {
+					return convertToMilliseconds(param.substring(valueStart + 1));
+				}
+			}
+		}
+		return null;
+	}
+
+	private static String parseStringParameter(String paramName, String[] parameters, String defaultValue) {
+		for (String param : parameters) {
+			if (param.startsWith(paramName + "=")) {
+				int valueStart = param.indexOf("=");
+				if (valueStart != -1) {
+					return param.substring(valueStart + 1);
+				}
+			}
+		}
+		return defaultValue;
+	}
+
+	private static DiagnosticProperties doJFR(String diagnosticCommand) {
+		DiagnosticProperties result = null;
+
+		// split the command and arguments
+		String[] parts = diagnosticCommand.split(DIAGNOSTICS_OPTION_SEPARATOR);
+		IPC.logMessage("doJFR: ", diagnosticCommand);
+
+		// ensure there's at least one part for the command
+		if (parts.length == 0) {
+			return DiagnosticProperties.makeErrorProperties("Error: No JFR command specified");
+		}
+
+		String command = parts[0].trim();
+		String[] parameters = Arrays.copyOfRange(parts, 1, parts.length);
+
+		String fileName = parseStringParameter("filename", parameters, null);
+		IPC.logMessage("doJFR: filename = ", fileName);
+		boolean setFileName = (fileName != null) && !fileName.isEmpty();
+		if (setFileName) {
+			if (!VM.setJFRRecordingFileName(fileName)) {
+				return DiagnosticProperties.makeErrorProperties("setJFRRecordingFileName failed");
+			} else {
+				jfrRecordingFileName = fileName;
+			}
+		}
+
+		try {
+			if (command.equalsIgnoreCase(DIAGNOSTICS_JFR_START)) {
+				VM.startJFR();
+				Long duration = parseTimeParameter("duration", parameters);
+				IPC.logMessage("doJFR: ", diagnosticCommand);
+				if (duration != null) {
+					long delay = duration.longValue();
+					if (delay > 0) {
+						Timer timer = new Timer();
+						TimerTask jfrDumpTask = new TimerTask() {
+							public void run() {
+								VM.jfrDump();
+							}
+						};
+						timer.schedule(jfrDumpTask, duration.longValue());
+					} else {
+						// the recording is on until next JFR command
+					}
+				}
+				result = DiagnosticProperties.makeStringResult("Start JFR recording to " + jfrRecordingFileName);
+			} else if (command.equalsIgnoreCase(DIAGNOSTICS_JFR_STOP)) {
+				VM.stopJFR();
```

Comment:
I think you should added another native `VM.isJFRRecordingStarted()` (the actual boolean can live in `vm->jfrState.isStarted`). If the recording hasnt started then it should output a message saying "Could not stop recording [jfrRecordingFileName]"

Reply:
Added `VM.isJFRRecordingStarted()` and native returning `vm->jfrState.isStarted`.
`JFR.dump()` & `JFR.stop()` requires `JFR.start()` first, otherwise an error message is printed.